runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/config/StreamingEngineComputationConfigFetcher.java

```diff
@@ -0,0 +1,296 @@
+  }
+
+  @Override
+  public Optional<ComputationConfig> getConfig(String computationId) {
+    Preconditions.checkArgument(
+        !computationId.isEmpty(),
+        "computationId is empty. Cannot fetch computation config without a computationId.");
+    return getComputationConfigInternal(computationId)
+        .flatMap(
+            config -> {
+              Map<String, String> stateNameMap = config.userStepToStateFamilyNameMap();
+              return config
+                  .computationConfig()
+                  .map(
+                      computationConfig ->
+                          ComputationConfig.create(
+                              createMapTask(computationConfig),
+                              computationConfig.getTransformUserNameToStateFamily(),
+                              stateNameMap));
+            });
+  }
+
+  @Override
+  public void stop() {
+    // We have already shutdown or start has not been called.
+    if (globalConfigRefresher.isShutdown() || !hasReceivedGlobalConfig.get()) {
+      return;
+    }
+
+    globalConfigRefresher.shutdown();
+    boolean isTerminated = false;
+    try {
+      isTerminated = globalConfigRefresher.awaitTermination(10, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      LOG.warn("Error occurred shutting down: {}", globalConfigRefresher);
+    }
+    if (!isTerminated) {
+      globalConfigRefresher.shutdownNow();
+    }
+  }
+
+  /**
+   * Initially b Schedules a background thread that periodically sends getConfig requests to
+   * Dataflow Service to obtain the windmill service endpoint.
+   */
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void schedulePeriodicGlobalConfigRequests() {
+    globalConfigRefresher.scheduleWithFixedDelay(
+        () -> getGlobalConfig().ifPresent(onStreamingConfig),
+        0,
+        globalConfigRefreshPeriodMillis,
+        TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Blocks until we have received the initial global config from Dataflow or an exception is
+   * thrown.
+   */
+  private synchronized void fetchInitialPipelineGlobalConfig() {
+    if (!hasReceivedGlobalConfig.get()) {
+      // Get the initial global configuration. This will initialize the windmillServer stub.
+      while (true) {
+        LOG.info("Sending request to get initial global configuration for this worker.");
+        Optional<StreamingEnginePipelineConfig> globalConfig = getGlobalConfig();
+        if (globalConfig.isPresent()) {
+          onStreamingConfig.accept(globalConfig.get());
+          hasReceivedGlobalConfig.compareAndSet(false, true);
+          break;
+        }
+        LOG.info("Haven't received initial global configuration, will retry in 5 seconds");
+        sleepUninterruptibly(5, TimeUnit.SECONDS);
+      }
+    }
+    LOG.info("Initial global configuration received, harness is now ready");
+  }
+
+  private Optional<StreamingEnginePipelineConfig> getComputationConfigInternal(
```

Comment:
I think this should just return the computationconfig not full StreamingEnginePipelineConfig.  I think that StreamingEnginePipelineConfig can have the computation config removed from it, as we don't care about listening to computations as they are fetched.

That will then make it clearer that we only want the listener to trigger on the periodic background fetching, not the computation fetching driven by the cache.

Reply:
done